tests-ng/DEVELOPER.md

```diff
@@ -0,0 +1,446 @@
+    # Test implementation
+```
+
+#### `@pytest.mark.root(reason="...")`
+
+Indicates the test requires root privileges:
+
+```python
+@pytest.mark.root(reason="Starting the unit requires root")
+def test_service_start(systemd: Systemd):
+    # Test implementation
+```
+
+#### `@pytest.mark.feature("condition", reason="...")`
+
+Limits test execution based on feature conditions:
+
+```python
+@pytest.mark.feature("ssh and not (ali or aws or azure or openstack)",
+                     reason="We want no authorized_keys for unmanaged users")
+def test_users_have_no_authorized_keys(expected_users):
+    # Test implementation
+```
+
+#### `@pytest.mark.performance_metric`
+
+Marks performance tests that can be skipped under emulation:
+
+```python
+@pytest.mark.performance_metric
+@pytest.mark.booted(reason="We can only measure startup time if we actually boot the system")
+def test_startup_time(systemd: Systemd):
+    # Test implementation
+```
+
+### Parametrization
+
+Use `@pytest.mark.parametrize` to avoid repetitive, nearly identical tests where the differences can naturally be expressed via parameters. However, don't overuse parametrization or strive for ultra-generic test functions at the expense of clarity. If extracting a parameter makes the test clearer and avoids straightforward duplication, use it. But a small amount of code duplication is perfectly acceptable in tests if it keeps things readable and concrete.
+
+**Yes – good use of parametrization:**
+
+```python
+@pytest.mark.parametrize("username", ["alice", "bob", "peter"])
+def test_home_directory_exists(username):
+    home = Path(f"/home/{username}")
+    assert home.exists(), f"Missing home for {username}"
+```
+
+**No – avoid excessive abstraction for unclear wins:**
+
+```python
+# This is too abstract—each "case" has different logic, result, or context.
+@pytest.mark.parametrize("input_val,expected", [(1, True), ("foo", False), ([], Exception)])
+def test_weird_cases(input_val, expected):
+    # ...too much unrelated logic for one test
+```
+
+**Guideline:** Prefer clarity and intent over DRY-ness in tests. Parametrize when it makes tests simpler, not just shorter.
+
+### Missing Markers (TODO)
+
+#### `@pytest.mark.security_id`
+
+TODO: Explain what this marker is good for.
+
+## Python Best Practices
+
+### Code Style and CI Enforcement
+
+The project enforces code quality through CI linting (see `.github/workflows/lint_tests-ng.yml`):
+
+- **[Black](https://black.readthedocs.io/en/stable/)**: Automatic code formatting
+- **[isort](https://pycqa.github.io/isort/)**: Import statement sorting
+- **[Pyright](https://github.com/microsoft/pyright/blob/main/docs/getting-started.md)**: Static type checking
+
+**Guidelines:**
+
+- Follow [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/)
+- Use type hints for function parameters and return values
+- Prefer descriptive variable names
+- Use docstrings for complex functions
```

Comment:
I think this wording is too lenient here. There are countless reasons why we should follow PEP 8 and python best practices as much as possible. So I would like to rephrase this more strongly.


```suggestion
1. All code contributions must follow the [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/)
2. The use of type hints is mandatory for all new additions to avoid typing related bugs and increase readability
3. Use descriptive variable and function names
4. Use docstrings to summarize functions and classes. The extent of which depends on the size, scope and complexity of the function or class.
5. Initialize all variables with a sensible default value instead of using `None` (unless the value and its type hint explicitly make it noneable)
```